File: code/backend/java/core/src/main/java/games/paths/core/service/match/RandomEventSelectionService.java
package games.paths.core.service.match;

import games.paths.core.model.match.MatchStatuses;
import games.paths.core.port.match.RandomEventStorePort;
import games.paths.core.port.match.RandomEventStorePort.RandomEventMatchContext;
import games.paths.core.port.match.RandomEventStorePort.RandomEventRuleView;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Random;
import java.util.Set;

/**
 * RandomEventSelectionService - picks at most one global random event per time-start (Step 39).
 * {@code probability} is an absolute percentage; a total above 100 scales every row down.
 */
public class RandomEventSelectionService {

    /** Keeps the random roll of day N away from the weather roll of day N+1 (seed + clock). */
    public static final long SEED_SALT = 1_000_003L;
    static final int PERCENT = 100;

    private final RandomEventStorePort store;
    private final RegistryService registryService;

    public RandomEventSelectionService(RandomEventStorePort store, RegistryService registryService) {
        this.store = store;
        this.registryService = registryService;
    }

    /** The event to fire at this time-start, or empty when nothing fires. */
    public Optional<RandomEventPick> pickAtTimeStart(long idMatch) {
        RandomEventMatchContext ctx = store.loadContext(idMatch).orElse(null);
        if (ctx == null || !MatchStatuses.RUNNING.equals(ctx.status()) || ctx.currentClock() <= 0) {
            return Optional.empty();
        }
        List<RandomEventRuleView> rows = store.findRandomEvents(ctx.idStory());
        if (rows == null || rows.isEmpty()) {
            return Optional.empty();
        }
        List<RandomEventRuleView> eligible = filterEligible(rows, idMatch, store.findConsumedEventIds(idMatch));
        return pick(eligible, seedFor(ctx))
                .map(r -> new RandomEventPick(r.id(), r.idEvent()));
    }

    private List<RandomEventRuleView> filterEligible(List<RandomEventRuleView> rows, long idMatch,
                                                     Set<Long> consumed) {
        List<RandomEventRuleView> out = new ArrayList<>();
        for (RandomEventRuleView r : rows) {
            if (isRunnable(r, consumed) && conditionMatches(r, idMatch)) {
                out.add(r);
            }
        }
        return out;
    }

    /** Positive probability, existing event without choices, and not a spent ONCE event. */
    static boolean isRunnable(RandomEventRuleView r, Set<Long> consumed) {
        if (r.probability() <= 0 || r.idEvent() == null || r.idEvent() <= 0
                || !r.eventExists() || r.eventOwnsChoices()) {
            return false;
        }
        boolean once = EventAvailabilityChecker.TYPE_ONCE.equalsIgnoreCase(r.eventType());
        return !(once && consumed != null && consumed.contains(r.idEvent().longValue()));
    }

    private boolean conditionMatches(RandomEventRuleView r, long idMatch) {
        if (RegistryService.noCondition(r.conditionKey())) {
            return true;
        }
        return RegistryService.evaluate(r.conditionOperator(), r.conditionValue(),
                registryService.find(idMatch, r.conditionKey()));
    }

    /** Probability DESC, then id ASC: the fixed order of the cumulative walk. */
    static List<RandomEventRuleView> order(List<RandomEventRuleView> eligible) {
        List<RandomEventRuleView> ordered = new ArrayList<>(eligible);
        ordered.sort(Comparator.comparingInt(RandomEventRuleView::probability).reversed()
                .thenComparingLong(RandomEventRuleView::id));
        return ordered;
    }

    /** The row whose cumulative range holds {@code roll}; empty when roll lands past the total. */
    static Optional<RandomEventRuleView> walk(List<RandomEventRuleView> ordered, int roll) {
        int cumulative = 0;
        for (RandomEventRuleView r : ordered) {
            cumulative += r.probability();
            if (roll < cumulative) {
                return Optional.of(r);
            }
        }
        return Optional.empty();
    }

    /** Roll in [0, max(100, total)): below 100 the gap is "nothing", above it the rows scale. */
    static Optional<RandomEventRuleView> pick(List<RandomEventRuleView> eligible, long seed) {
        if (eligible == null || eligible.isEmpty()) {
            return Optional.empty();
        }
        return walk(order(eligible), new Random(seed).nextInt(Math.max(PERCENT, total(eligible))));
    }

    static int total(List<RandomEventRuleView> eligible) {
        int total = 0;
        for (RandomEventRuleView r : eligible) {
            total += r.probability();
        }
        return total;
    }

    static long seedFor(RandomEventMatchContext ctx) {
        long base = ctx.rngSeed() != null ? ctx.rngSeed() : ctx.idStory();
        return base + ctx.currentClock() + SEED_SALT;
    }

    /** The row that fired and the event it runs. */
    public record RandomEventPick(long idRandomEvent, long idEvent) {
    }
}

File: code/backend/python/app/core/services/match/random_event_selection_service.py
"""Step 39 — picks at most one global random event per time-start (mirrors the Java reference).
``probability`` is an absolute percentage; a total above 100 scales every row down.
"""
import random
from typing import Any, Dict, List, Optional, Set

from app.core.models.match import match_statuses
from app.core.services.match import registry_service

# Keeps the random roll of day N away from the weather roll of day N+1 (seed + clock).
SEED_SALT = 1_000_003
PERCENT = 100
TYPE_ONCE = "ONCE"


class RandomEventSelectionService:
    def __init__(self, store, registry_service_instance=None) -> None:
        self.store = store
        self.registry_service = registry_service_instance

    def pick_at_time_start(self, id_match: int) -> Optional[Dict[str, int]]:
        """The event to fire at this time-start as ``{"id_random_event", "id_event"}``, or None."""
        ctx = self.store.load_context(id_match)
        if (ctx is None or ctx.get("status") != match_statuses.RUNNING
                or (ctx.get("current_clock") or 0) <= 0):
            return None
        rows = self.store.find_random_events(ctx["id_story"]) or []
        if not rows:
            return None
        consumed = self.store.find_consumed_event_ids(id_match)
        eligible = [r for r in rows
                    if self.is_runnable(r, consumed) and self._condition_matches(r, id_match)]
        chosen = self.pick(eligible, self.seed_for(ctx))
        if chosen is None:
            return None
        return {"id_random_event": chosen["id"], "id_event": chosen["id_event"]}

    @staticmethod
    def is_runnable(row: Dict[str, Any], consumed: Optional[Set[int]]) -> bool:
        """Positive probability, existing event without choices, and not a spent ONCE event."""
        if (_probability(row) <= 0 or not row.get("id_event") or row["id_event"] <= 0
                or not row.get("event_exists") or row.get("event_owns_choices")):
            return False
        once = (row.get("event_type") or "").upper() == TYPE_ONCE
        return not (once and consumed and row["id_event"] in consumed)

    def _condition_matches(self, row: Dict[str, Any], id_match: int) -> bool:
        key = row.get("condition_key")
        if registry_service.no_condition(key):
            return True
        return registry_service.evaluate(row.get("registry_value_operator_condition"),
                                         row.get("condition_value"),
                                         self.registry_service.find(id_match, key))

    @staticmethod
    def order(eligible: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Probability DESC, then id ASC: the fixed order of the cumulative walk."""
        return sorted(eligible, key=lambda r: (-_probability(r), r["id"]))

    @staticmethod
    def walk(ordered: List[Dict[str, Any]], roll: int) -> Optional[Dict[str, Any]]:
        """The row whose cumulative range holds ``roll``; None when roll lands past the total."""
        cumulative = 0
        for r in ordered:
            cumulative += _probability(r)
            if roll < cumulative:
                return r
        return None

    @classmethod
    def pick(cls, eligible: List[Dict[str, Any]], seed: int) -> Optional[Dict[str, Any]]:
        """Roll in [0, max(100, total)): below 100 the gap is "nothing", above it the rows scale."""
        if not eligible:
            return None
        # Safe: seed is provided externally (deterministic, not for security).
        roll = random.Random(int(seed)).randrange(max(PERCENT, cls.total(eligible)))  # NOSONAR
        return cls.walk(cls.order(eligible), roll)

    @staticmethod
    def total(eligible: List[Dict[str, Any]]) -> int:
        return sum(_probability(r) for r in eligible)

    @staticmethod
    def seed_for(ctx: Dict[str, Any]) -> int:
        base = ctx.get("rng_seed") if ctx.get("rng_seed") is not None else ctx["id_story"]
        return int(base) + int(ctx.get("current_clock") or 0) + SEED_SALT


def _probability(row: Dict[str, Any]) -> int:
    return int(row.get("probability") or 0)
